File: week-04/main.cpp
#include <cstdio>
#include <chrono>
#include <cstdint>

static inline double calculate_optimized(uint32_t iterations) {
    double result = 1.0;
    double d1 = 3.0;  // 4*i - 1 starting at i=1
    double d2 = 5.0;  // 4*i + 1 starting at i=1

    uint32_t i = 0;
    uint32_t n = iterations;
    uint32_t limit = n & ~15u; // multiple of 16

    for (; i < limit; i += 16) {
        result -= 1.0 / d1; result += 1.0 / d2; d1 += 4.0; d2 += 4.0;
        result -= 1.0 / d1; result += 1.0 / d2; d1 += 4.0; d2 += 4.0;
        result -= 1.0 / d1; result += 1.0 / d2; d1 += 4.0; d2 += 4.0;
        result -= 1.0 / d1; result += 1.0 / d2; d1 += 4.0; d2 += 4.0;

        result -= 1.0 / d1; result += 1.0 / d2; d1 += 4.0; d2 += 4.0;
        result -= 1.0 / d1; result += 1.0 / d2; d1 += 4.0; d2 += 4.0;
        result -= 1.0 / d1; result += 1.0 / d2; d1 += 4.0; d2 += 4.0;
        result -= 1.0 / d1; result += 1.0 / d2; d1 += 4.0; d2 += 4.0;

        result -= 1.0 / d1; result += 1.0 / d2; d1 += 4.0; d2 += 4.0;
        result -= 1.0 / d1; result += 1.0 / d2; d1 += 4.0; d2 += 4.0;
        result -= 1.0 / d1; result += 1.0 / d2; d1 += 4.0; d2 += 4.0;
        result -= 1.0 / d1; result += 1.0 / d2; d1 += 4.0; d2 += 4.0;

        result -= 1.0 / d1; result += 1.0 / d2; d1 += 4.0; d2 += 4.0;
        result -= 1.0 / d1; result += 1.0 / d2; d1 += 4.0; d2 += 4.0;
        result -= 1.0 / d1; result += 1.0 / d2; d1 += 4.0; d2 += 4.0;
        result -= 1.0 / d1; result += 1.0 / d2; d1 += 4.0; d2 += 4.0;
    }

    for (; i < n; ++i) {
        result -= 1.0 / d1;
        result += 1.0 / d2;
        d1 += 4.0;
        d2 += 4.0;
    }

    return result;
}

int main() {
    const uint32_t iterations = 200000000u;
    const int param1 = 4;
    const int param2 = 1;

    auto start_time = std::chrono::steady_clock::now();
    double result = calculate_optimized(iterations) * 4.0;
    auto end_time = std::chrono::steady_clock::now();

    double elapsed = std::chrono::duration<double>(end_time - start_time).count();

    std::printf("Result: %.12f\n", result);
    std::printf("Execution Time: %.6f seconds\n", elapsed);
    return 0;
}

File: week-04/main.rs
use std::time::Instant;

const A: u32 = 1664525;
const C: u32 = 1013904223;

#[inline(always)]
fn lcg_next(state: &mut u32) -> u32 {
    *state = state.wrapping_mul(A).wrapping_add(C);
    *state
}

#[inline(always)]
fn max_subarray_sum(n: usize, seed: u32, min_val: i128, max_val: i128) -> i128 {
    let mut state = seed;
    let range = max_val - min_val + 1;
    // Assume valid input where range > 0
    let use_u32 = range <= (u32::MAX as i128);
    let r32 = if use_u32 { range as u32 } else { 0 };
    let r128 = if use_u32 { 0 } else { range as u128 };

    // Kadane's algorithm
    let mut best: i128 = i128::MIN;
    let mut curr: i128 = 0;

    let mut i = 0;
    while i < n {
        let v = lcg_next(&mut state);
        let rem_i128: i128 = if use_u32 {
            (v % r32) as i128
        } else {
            ((v as u128) % r128) as i128
        };
        let x = rem_i128 + min_val;

        if curr > 0 {
            curr += x;
        } else {
            curr = x;
        }
        if curr > best {
            best = curr;
        }

        i += 1;
    }
    best
}

#[inline(always)]
fn total_max_subarray_sum(n: usize, initial_seed: u32, min_val: i128, max_val: i128) -> i128 {
    let mut total: i128 = 0;
    let mut gen_state = initial_seed;
    let mut k = 0;
    while k < 20 {
        let seed = lcg_next(&mut gen_state);
        total += max_subarray_sum(n, seed, min_val, max_val);
        k += 1;
    }
    total
}

fn main() {
    let n: usize = 10000;
    let initial_seed: u32 = 42;
    let min_val: i128 = -10;
    let max_val: i128 = 10;

    let start = Instant::now();
    let result = total_max_subarray_sum(n, initial_seed, min_val, max_val);
    let elapsed = start.elapsed().as_secs_f64();

    println!("Total Maximum Subarray Sum (20 runs): {}", result);
    println!("Execution Time: {:.6} seconds", elapsed);
}